### PR TITLE
Remove support for Gopher protocol

### PR DESCRIPTION
**Pull request by rousskov**
Gopher code quality remains too low for production use in most
environments. The code is a persistent source of vulnerabilities and
fixing it requires significant effort. We should not be spending scarce
Project resources on improving that code, especially given the lack of
strong demand for Gopher support.

With this change, Gopher requests will be handled like any other request
with an unknown (to Squid) protocol. For example, HTTP requests with
Gopher URI scheme result in ERR_UNSUP_REQ.

Default Squid configuration still considers TCP port 70 "safe". The
corresponding Safe_ports ACL rule has not been removed for consistency
sake: We consider WAIS port safe even though Squid refuses to forward
WAIS requests:

    acl Safe_ports port 70          # gopher
    acl Safe_ports port 210         # wais

### CODE REVIEW

File: doc/Programming-Guide/Groups.dox [hunk 1]
```diff
@@ -33,7 +33,7 @@
  *   All requests to origin-server are sent in that servers protocol.
  *
  \par
- *   Wais and Gopher don't receive much
+ *   Wais and Whois don't receive much
```
Comment (yadij) [inline doc/Programming-Guide/Groups.dox:36]: This is wrong. Whois is an ACL lookup and still frequently used by major proxies.

File: errors/az.po [hunk 2]
```diff
@@ -845,7 +845,7 @@ msgid ""
 "Squid does not support some access protocols. For example, the SSH protocol "
 "is currently not supported."
 msgstr ""
-"Squid bütün sorğu metodları və bütün protokollardəstəkləmir. Məsələn, Gopher "
+"Squid bütün sorğu metodları və bütün protokollardəstəkləmir. Məsələn, WHOIS "
```
[the remark this answers is not in the document]
Comment (yadij) [reply errors/az.po:848]: Please do not edit PO/POT msgid strings directly. They are automatically generated from the src/ manuals and error/templates/ files. Changing those original source texts is sufficient for a code/documentation PR.

**Comment by yadij** [comment]
> Gopher code quality remains too low for production use in most
environments.

Subjective opinion. Most of Squid HTTP/1 code is lower quality than this Gopher code. Shall we erase HTTP/1 support?

> The code is a persistent source of vulnerabilities and
fixing it requires significant effort.

False. This code has far lower incidence of vulnerabilities than the other protocols still supported (HTTP , FTP).
The known issues with gopher are small and code is fixed relatively easily when developer time is applied[1]. HTTP/FTP by comparison have major design vulnerabilities in the protocols themselves - thus we will never be able to fix those others.

[1] the major reason this gopher code being so old as it is (and one vulnerability still existing) has been your own objections to others to updating the code in recent years. Drive for perfection has killed progress.


> We should not be spending scarce
Project resources on improving that code, especially given the lack of
strong demand for Gopher support.

Very corporate approach there. "strength" of demand is not very relevant in FOSS like Squid. It is also a subjective immeasurable criteria.

 * We still have both developers, and users of gopher protocol in the Squid community.

 * There are a (growing!) number of gopher services available on the public Internet and Squid is one of the few pieces of software which can provide gateway functionality. This protocol is still useful.

**Review by yadij — CHANGES_REQUESTED** [review]
Notwithstanding my basic objection to the removal. There are issues that need to be addressed with the PR change, marked below.

File: errors/templates/ERR_UNSUP_REQ [hunk 3]
```diff
@@ -24,7 +24,7 @@ body
 <p><b>Unsupported Request Method and Protocol</b></p>
 </blockquote>
 
-<p>Squid does not support all request methods for all access protocols. For example, you can not POST a Gopher request.</p>
+<p>Squid does not support all request methods for all access protocols. For example, you can not POST a WHOIS request.</p>
```
[the remark this answers is not in the document]
Comment (rousskov) [reply errors/templates/ERR_UNSUP_REQ]: The example removed in cdc07d4.

**Review by rousskov — COMMENTED** [review]
> Notwithstanding my basic objection to the removal.

FWIW, I am not sure _why_ you are objecting. The PR provides explicit/specific reasons for Gopher removal. Your comments did not invalidate those reasons AFAICT. Your reasoning for keeping Gopher can also be valid, of course, but it is currently unknown (to me). If I extract bits and pieces from your comments that focus on _PR_ reasons, I am sure I will not accurately reconstruct _your_ reasoning for keeping Gopher. It would be nice if you state it explicitly instead.

> There are issues that need to be addressed with the PR change, marked below.

I hope all issues are addressed (as of commit cdc07d4).

**Comment by rousskov** [comment]
> given it's state of maintenance, possibly it should be disabled by default at least? I expect you have considered this already, though.

Correctly making Gopher support conditional requires more work than removing it, makes the high-level code logic more complex, and does not address any of the Gopher code maintenance problems. And turning that support off by default makes any maintenance even less likely to occur. The Project may still decide to spend precious resources on that, of course, but I think a clean removal is an overall better solution.

Needless to say, those who really need Gopher support can maintain an unofficial removal reversal (i.e. Gopher addition) patch. That would be a much better allocation of Project resources IMO. Other options include developing plugins (i.e. loadable modules or adapters) that add optional Gopher support. Overall, just because there is no Gopher code in the official Squid distribution does not mean those who need Gopher support cannot make it happen, of course.

File: doc/Programming-Guide/Groups.dox (hunk 1 above)
Comment (yadij) [reply doc/Programming-Guide/Groups.dox]: Yes, thank you.

**Review by yadij — COMMENTED** [review]
I see no technical issues now. Removing my veto.

**Review by yadij — APPROVED** [review]

**Comment by rousskov** [comment]
@kinkie, what is your position regarding Gopher support removal?

**Pull request closed without merging**